tests/unit/test_charm.py

```diff
@@ -0,0 +1,160 @@
+            reason = "AlreadyExists"
+        return {
+            "apiVersion": 1,
+            "code": self.code,
+            "message": "broken",
+            "reason": reason,
+        }
+
+
+class _FakeApiError(ApiError):
+    """Used to simulate an ApiError during testing."""
+
+    def __init__(self, code=400):
+        super().__init__(response=_FakeResponse(code))
+
+
+@patch("lightkube.core.client.GenericSyncClient", Mock)
+class TestCharm(unittest.TestCase):
+    def setUp(self):
+        self.harness = Harness(TrainingOperatorCharm)
+        self.addCleanup(self.harness.cleanup)
+        self.harness.begin()
+
+    # Event handlers
+    @patch("charm.TrainingOperatorCharm._create_crds")
+    @patch("charm.TrainingOperatorCharm._create_resource")
+    def test_install_event(self, create_auth, create_crds):
+        self.harness.charm.on.install.emit()
+
+        # Ensure create_auth and create_crds
+        # are called once
+        create_auth.assert_called_once()
+        create_crds.assert_called_once()
+
+        # Check status is Active
+        self.assertEqual(self.harness.charm.unit.status, ActiveStatus())
+
+    @patch("charm.TrainingOperatorCharm._create_crds")
+    @patch("charm.TrainingOperatorCharm._create_resource")
+    def test_blocked_on_appierror_on_install_event(self, create_auth, create_crds):
+        # Ensure the unit is in BlockedStatus
+        # on exception when creating auth resources
+        subtests = (
+            (
+                create_auth,
+                "Test BlockedStatus when ApiError is raised trying to create auth resources",
+            ),
+            (
+                create_crds,
+                "Test BlockedStatus when ApiError is raised trying to create CRDS",
+            ),
+        )
+        for create_type, subtest_description in subtests:
+            with self.subTest(msg=subtest_description):
+                create_type.side_effect = _FakeApiError()
+                try:
+                    self.harness.charm.on.install.emit()
+                except ApiError:
+                    self.assertEqual(
+                        self.harness.charm.unit.status,
+                        BlockedStatus(
+                            f"Creating/patching resources failed with code {create_type.side_effect.response.code}."
+                        ),
+                    )
+
+    @patch("charm.TrainingOperatorCharm._update_layer")
+    @patch("charm.TrainingOperatorCharm._patch_resource")
+    def test_config_changed_event(self, patch, update):
+        self.harness.charm.on.config_changed.emit()
+
+        # Ensure _update_layer is called once and
+        # _patch_resource is called
+        update.assert_called_once()
+        patch.assert_called()
+
+        # Ensure the unit is in BlockedStatus
+        # on exception when patching auth resources
+        patch.side_effect = _FakeApiError()
+        try:
+            self.harness.charm.on.config_changed.emit()
```

Comment:
It looks like this is trying to test what happens on a failure. If so, we might want to use `assertRaises` here:

```python
with self.assertRaises(ApiError):
    self.harness.charm.on.config_changed.emit()
self.assertEqual(...)
```

Reply:
Yes, that was my first approach, unfortunately the code won't allow us to do that. I have this in my queue of improvements for `src/charm.py`.